pkg/boot/grub/echo_test.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017-2020 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package grub
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/u-root/u-root/pkg/curl"
+)
+
+var update = flag.Bool("run-bash", false, "run bash and update golden file")
+
+func TestMain(m *testing.M) {
+	if os.Getenv("BE_ECHO") == "1" {
+		fmt.Printf("echo:%#v\n", os.Args[1:])
+		return
+	} // call flag.Parse() here if TestMain uses flags
+	os.Exit(m.Run())
+}
+
+func TestHelperEcho(t *testing.T) {
+	cmd := exec.Command(os.Args[0], "echothis")
+	cmd.Env = append(os.Environ(), "BE_ECHO=1")
+	out, err := cmd.Output()
+	t.Logf("%q\n", out)
+	if err != nil {
+		t.Fatalf("process ran with err %v", err)
+	}
+	want := "echo:[]string{\"echothis\"}\n"
+	if string(out) != want {
+		t.Fatalf("wrong process output got `%s` want `%s`", out, want)
+	}
+}
+
+func TestBashWrapper(t *testing.T) {
+	if !*update {
+		t.Skip("use -run-bash flag to run this")
+	}
+	cmd := exec.Command("./testdata/bash_wrapper.sh", os.Args[0], "./testdata/test_bash_wrapper.sh")
+	cmd.Env = append(os.Environ(), "BE_ECHO=1")
+	out, err := cmd.Output()
+	t.Logf("%q\n", out)
+	if err != nil {
+		t.Fatalf("process ran with err %v", err)
+	}
+	want := "echo:[]string{\"param1\", \"param2\"}\n"
+	if string(out) != want {
+		t.Fatalf("wrong process output got `%s` want `%s`", out, want)
+	}
+}
+
+func TestGrubTests(t *testing.T) {
+	files, err := filepath.Glob("testdata/*.in")
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, file := range files {
+		name := strings.TrimSuffix(filepath.Base(file), ".in")
+		t.Run(name, func(t *testing.T) {
+			golden := strings.TrimSuffix(file, ".in") + ".out"
+			var out []byte
+			if *update {
+				cmd := exec.Command("./testdata/bash_wrapper.sh", os.Args[0], file)
```

Comment:
would you mind explaining the bash wrapper in comments? I don't understand what it or the golden file does

Reply:
yes, I'll do that, probably adapting the comment I added below